readme.md
# quantileplot

An R Package to visualize bivariate relationships with a quantile plot.

Install this package with the command `devtools::install_github("ilundberg/quantileplot")`. Then see the [vignette](https://ilundberg.github.io/quantileplot/doc/quantileplot.html) to get started.

**Warning:** This package is actively being developed. Functionality is likely to change in the near future. You should think of it as a directory of code that exists on Github. If you use any of this code (or the package as a whole), you should store a copy of the version you are using so that you can reproduce your results. The package as available here on Github will change. This warning will be updated when the package is stable.

## Sample Use
`quantileplot(y ~ s(x), data)`

```
x <- rbeta(1000,1,2)
y <- log(1 + 9 * x) * rbeta(1000, 1, 2)
data <- data.frame(x = x, y = y)
quantileplot(y ~ s(x), data)
```

![Output of the quantileplot function](example.png)


news.md
# NEWS file for the quantileplot package

# quantileplot 0.1.5
* Updated case_when() to coerce to numeric, since NA sometimes was interpreted as logical
* Added check that x- and y-variables are numeric
* Added coercion to numeric (with warning) to address the setting where x or y were haven_labelled
* This development version is available on GitHub.
* This is a beta version and the API may change before submission to CRAN.

# quantileplot 0.1.4
* This development version is available on GitHub.
* This is a beta version and the API may change before submission to CRAN.


tests/testthat.R
library(testthat)
library(quantileplot)

test_check("quantileplot")


tests/testthat/test-quantileplot-output.R
library(quantileplot)

testthat::test_that("Test functionality with a basic call", {
  test_data <- data.frame(x = runif(100),
                          y = runif(100))
  test <- quantileplot(y ~ s(x),
                       data = test_data)
  testthat::expect_true(inherits(test,"quantileplot"))
  testthat::expect_true(inherits(test$plot,"ggplot"))
  testthat::expect_true(inherits(test$mqgam.out,"mqgam"))
  testthat::expect_equal(length(test$plot$layers),6)
})

testthat::test_that("Test functionality with confidence bands", {
  test_data <- data.frame(x = runif(100),
                          y = runif(100))
  test <- quantileplot(y ~ s(x),
                       data = test_data,
                       show_ci = T)
  testthat::expect_true(inherits(test,"quantileplot"))
  testthat::expect_true(inherits(test$plot,"ggplot"))
  testthat::expect_equal(length(test$plot$layers),7)
})

testthat::test_that("Test functionality with truncated x_data_range and y_data_range", {
  test_data <- data.frame(x = runif(100),
                          y = runif(100))
  test <- quantileplot(y ~ s(x),
                       data = test_data,
                       x_data_range = c(.2,.8),
                       y_data_range = c(.3,.7))
  testthat::expect_true(inherits(test,"quantileplot"))
  testthat::expect_true(inherits(test$plot,"ggplot"))
  testthat::expect_equal(test$x_data_range, c(.2,.8))
  testthat::expect_equal(test$y_data_range, c(.3,.7))
})

testthat::test_that("Test using labels instead of a legend", {
  test_data <- data.frame(x = runif(100),
                          y = runif(100))
  test <- quantileplot(y ~ s(x),
                       data = test_data,
                       quantile_notation = "label")
  testthat::expect_true(inherits(test,"quantileplot"))
  testthat::expect_true(inherits(test$plot,"ggplot"))
})

testthat::test_that("Test with weights", {
  test_data <- data.frame(x = runif(100),
                          y = runif(100),
                          weight = runif(100, .4,.6))
  test <- quantileplot(y ~ s(x),
                       data = test_data,
                       weights = "weight")
  testthat::expect_true(inherits(test,"quantileplot"))
  testthat::expect_true(inherits(test$plot,"ggplot"))
})


tests/testthat/test-quantileplot-argument-errors.R
library(quantileplot)

testthat::test_that("Test that incorrect arguments throw errors", {
  test_data <- data.frame(x = runif(100),
                          y = runif(100),
                          z = runif(100))
  # Error if you give more than one predictor
  testthat::expect_error(quantileplot(y ~ s(x) + z, data = test_data))
  # Error if you don't give data
  testthat::expect_error(quantileplot(y ~ s(x)))
  # Error if you give weights that are not a character
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, weights = c(1,2,3)))
  # Error if weights are not a column of the data
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, weights = "nonexistent_column"))
  # Error if quantiles outside (0,1)
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, q = -2))
  # Error if xlab or ylab is something that does not belong on an axis title, like a function
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, xlab = mean))
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, ylab = mean))
  # Error if x_data_range or y_data_range is not a vector of length 2
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, x_data_range = 1))
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, y_data_range = 1))
  # Error if x_breaks or y_breaks is something other than a function or numeric vector
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, x_breaks = c("a","b")))
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, y_breaks = c("a","b")))
  # Error if x_labels or y_labels given without breaks specified
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, x_labels = c("a","b")))
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, y_labels = c("a","b")))
  # Error if x_labels (y_labels) does not match the length of x_breaks (y_breaks)
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, x_breaks = c(.2,.5,.7), x_labels = c("a","b")))
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, y_breaks = c(.2,.5,.7), y_labels = c("a","b")))
  # Error if x_bw or y_bw is something other than a numeric scalar
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, x_bw = c(.2,.5,.7), x_labels = c("a","b")))
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, x_bw = "string"))
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, y_bw = c(.2,.5,.7), x_labels = c("a","b")))
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, y_bw = "string"))
  # Error if quantile_notation is some unknown string
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, quantile_notation = "not_an_option"))
  # Error if credibility level is not a numeric scalar
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, credibility_level = c(.5,.95)))
  # Error if granularity is not a whole number scalar
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, granularity = c(100,200)))
  # Error if argGam is not a list
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, argGam = "method"))
  # Error if previous_fit is not a quantileplot object
  testthat::expect_error(quantileplot(y ~ s(x), data = test_data, previous_fit = "not_a_quantileplot"))


  test <- quantileplot(y ~ s(x),
                       data = test_data,
                       quantile_notation = "label")
})


R/gen_densities.R
#' Generate Density Estimates for a \code{quantileplot}
#'
#' @description Estimate the conditional density of the outcome at at particular values of the predictor. These conditional densities appear as vertical slices in the output of \code{\link{quantileplot}}. This function is typically called indirectly via a user call to \code{\link{quantileplot}}.
#'
#' @param data Data frame containing columns \code{x}, \code{y}, and \code{weight}. For a simple random sample, set \code{weight} to be constant.
#' @param slice_n Integer number of vertical slices (conditional densities of y given x) to be plotted. Default is 5.
#' @param x_data_range Numeric vector of length 2 containing the range of horizontal values to be plotted. Defaults to the range of the predictor variable in \code{data}. You may want to specify a narrower range if the predictor is extremely skewed. Quantile curves and densities will be estimated only on data in this range, and the plot will note the percent truncated.
#' @param y_data_range Numeric vector of length 2 containing the range of vertical values to be plotted. Defaults to the range of the outcome variable in \code{data}. You may want to specify a narrower range if the outcome is extremely skewed. Densities are truncated to this range. All data contribute to quantile curve estimation regardless of \code{y_data_range} to avoid selection on the outcome, though the visualization is truncated to \code{y_data_range}. The plot will note the percent truncated.
#' @param x_bw Numeric bandwidth for density estimation in the \code{x} dimension. The standard deviation of a Gaussian kernel. If \code{NULL}, this is set by the defaults in \code{stats::density()}.
#' @param y_bw Numeric bandwidth for density estimation in the \code{y} dimension. The standard deviation of a Gaussian kernel. If \code{NULL}, this is set by the defaults in \code{stats::density()}.
#' @param granularity Integer number of points at which to evaluate each density. Defaults to 512, as in \code{stats::density()}. Higher values yield more granular density estimates.
#'
#' @return List containing \code{marginal} and \code{conditional}, each of which is a data frame with density estimates.
#'
#' @references Lundberg, Ian, Robin C. Lee, and Brandon M. Stewart. 2021. "The quantile plot: A visualization for bivariate population relationships." Working paper.
#' @references Lundberg, Ian, and Brandon M. Stewart. 2020. "Comment: Summarizing income mobility with multiple smooth quantiles instead of parameterized means." Sociological Methodology 50(1):96-111.
#' @references Fasiolo, Matteo, Simon N. Wood, Margaux Zaffran, Raphaël Nedellec, and Yannig Goude. 2020. "Fast calibrated additive quantile regression." Journal of the American Statistical Association.
#'
#' @export
#' @import dplyr
gen_densities <- function(data, slice_n = 7, x_data_range = NULL, y_data_range = NULL, x_bw = NULL, y_bw = NULL, granularity = 512) {
  if (!all(c("x","y","weight") %in% colnames(data))) {
    stop("The data object passed to gen_densities must have these columns: x, y, weight")
  }
  if (any(is.na(data$x))) {
    stop("In the data object passed to gen_densities, x has missing values")
  }
  if (any(is.na(data$y))) {
    stop("In the data object passed to gen_densities, y has missing values")
  }
  if (any(is.na(data$weight))) {
    stop("In the data object passed to gen_densities, weight has missing values")
  }
  # Normalize the weight as requested by stats::density()
  data$weight <- data$weight / sum(data$weight)
  # If bandwidths were not specified by the user, set them by the default from stats::density()
  if (is.null(x_bw) | is.null(y_bw)) {
    cat("Bandwidths are being set at:\n")
  }
  if (is.null(x_bw)) {
    x_bw <- stats::density(data$x, weights = data$weight, kernel = "gaussian")$bw
    cat(paste("x_bw =",x_bw,"\n"))
  }
  if (is.null(y_bw)) {
    y_bw <- stats::density(data$y, weights = data$weight, kernel = "gaussian")$bw
    cat(paste("y_bw =",y_bw,"\n"))
  }
  # If x_data_range or y_data_range is null, replace with the range of the data
  if (is.null(x_data_range)) {
    x_data_range <- range(data$x)
  }
  if (is.null(y_data_range)) {
    y_data_range = range(data$y)
  }

  # Define the grid of values at which to estimate densities.
  # Spread the grid evenly over the range, excluding the endpoints.
  marginal_x_seq <- seq(x_data_range[1], x_data_range[2], length.out = granularity + 2)[2:(granularity + 1)]
  # Use labeling::extended to select slice locations to match the vertical major grid lines
  grid_line_locations <- labeling::extended(x_data_range[1], x_data_range[2], slice_n + 2)
  # If that produced the correct number and the far right one has enough room before the
  # edge of the plot, then use those values
  if (length(grid_line_locations) == slice_n + 2 &
      x_data_range[2] - max(grid_line_locations) >= diff(grid_line_locations)[1]) {
    conditional_x_seq <- grid_line_locations[2:(slice_n + 1)]
  } else {
    # Sometimes labeling::extended chooses a different number of points to help with label making.
    # In that case, stick with the number of slices the user requested.
    conditional_x_seq <- seq(x_data_range[1], x_data_range[2], length.out = slice_n + 2)[2:(slice_n + 1)]
  }
  rm(grid_line_locations)
  # For the conditional y sequence, manually set a sequence based on the chosen granularity
  conditional_y_seq <- seq(y_data_range[1], y_data_range[2], length.out = granularity + 2)[2:(granularity + 1)]

  # Estimate the marginal density
  marginal_density <- data.frame(x = rep(NA, granularity),
                                 density = rep(NA, granularity))
  for (i in 1:granularity) {
    x_eval <- marginal_x_seq[i]
    contribution_each_observation <- 1 / x_bw * stats::dnorm((x_eval - data$x) / x_bw)
    estimate <- stats::weighted.mean(contribution_each_observation,
                                     w = data$weight)
    # Place those in the marignal_density data frame
    marginal_density$x[i] <- x_eval
    marginal_density$density[i] <- estimate
  }

  # Estimate the conditional densities
  conditional_density <- data.frame(x = rep(conditional_x_seq, each = granularity),
                                    y = rep(conditional_y_seq, slice_n),
                                    density = NA)
  for (i in 1:nrow(conditional_density)) {
    # For easier reference, extract this (x,y) for evaluation
    x_eval <- conditional_density$x[i]
    y_eval <- conditional_density$y[i]

    # Estimate contribution of each data point to the marginal density at (x_eval)
    m_each <- 1 / x_bw * stats::dnorm((x_eval - data$x) / x_bw)
    # Average those to produce the marginal density estimate at (x_eval)
    # Do the weighted mean manually to reduce risk of underflow problems in large samples.
    m <- sum(m_each * data$weight) / sum(data$weight)

    # Estimate contribution of each data point to the conditional density at (x_eval,y_eval)
    # using the product kernel.
    # Divide by m internal to the weighted mean rather than external so that the internal
    # product is not so small.
    # This is similar to Hyndman et al. 1996 eq. 2.4-2.5.
    g_each <- (m_each / m) * 1 / y_bw * stats::dnorm((y_eval - data$y) / y_bw)
    # Average those to produce the marginal density estimate at (x_eval, y_eval)
    g <- sum(g_each * data$weight) / sum(data$weight)

    # Place that in the data frame
    conditional_density$density[i] <- g
  }
  return(list(marginal = marginal_density,
              conditional = conditional_density,
              x_bw = x_bw,
              y_bw = y_bw))
}



R/plot.quantileplot.R
#' Plot function for \code{quantileplot} objects
#' @description Plots a \code{quantileplot} object.
#' @param x An object of class \code{quantileplot}, which results from a call to the function \code{quantileplot}
#' @param bottom_xlab_angle Numeric degrees to rotate x-axis labels
#' @param bottom_height Proportion of vertical space in the full plot to be taken up by uncertainty draws at the bottom (only relevant if the \code{quantileplot} includes uncertainty draws).
#' @param ... Other arguments to \code{plot} commands
#' @return Prints a plot. If the \code{quantileplot} was specified with option \code{uncertainty_draws = TRUE}, this is a \code{gtable} object with a panel of plots. If the \code{quantileplot} was specified with \code{uncertainty_draws = FALSE} (the default), this is a \code{ggplot2} object.
#' @references Lundberg, Ian, Robin C. Lee, and Brandon M. Stewart. 2021. "The quantile plot: A visualization for bivariate population relationships." Working paper.
#' @references Lundberg, Ian, and Brandon M. Stewart. 2020. "Comment: Summarizing income mobility with multiple smooth quantiles instead of parameterized means." Sociological Methodology 50(1):96-111.
#' @references Fasiolo, Matteo, Simon N. Wood, Margaux Zaffran, Raphaël Nedellec, and Yannig Goude. 2020. "Fast calibrated additive quantile regression." Journal of the American Statistical Association.
#' @export

plot.quantileplot <- function(x, bottom_xlab_angle = NULL, bottom_height = .3, ...) {
  if (is.null(x$arguments$uncertainty_draws)) {
    print(x$plot)
  } else {
    if (!is.numeric(bottom_height)) {
      stop("Argument error: Bottom height should be between 0 and 1 (a proportion of total plot height)")
    }
    if (bottom_height < 0 | bottom_height > 1) {
      stop("Argument error: Bottom height should be between 0 and 1 (a proportion of total plot height)")
    }
    all_plots <- x$sim_curve_plots
    if (!is.null(bottom_xlab_angle)) {
      for (i in 1:length(all_plots)) {
        all_plots[[i]] <- all_plots[[i]] +
          ggplot2::theme(axis.text.x = ggplot2::element_text(angle = bottom_xlab_angle, hjust = 1))
      }
    }
    all_plots$point <- x$plot
    with_uncertainty_draws <- gridExtra::grid.arrange(grobs = all_plots,
                                                      heights = c(1 - bottom_height, bottom_height),
                                                      layout_matrix = rbind(length(x$sim_curve_plots) + 1,
                                                                            1:length(x$sim_curve_plots)))

    print(with_uncertainty_draws)
  }
}


R/print.quantileplot.R
#' Print function for \code{quantileplot} objects
#' @description Prints the plot generated by \code{quantileplot}
#' @param x An object of class \code{quantileplot}, which results from a call to the function \code{quantileplot}
#' @param ... Other arguments to \code{print} commands
#' @return Prints the plot generated by \code{quantileplot}
#' @references Lundberg, Ian, Robin C. Lee, and Brandon M. Stewart. 2021. "The quantile plot: A visualization for bivariate population relationships." Working paper.
#' @references Lundberg, Ian, and Brandon M. Stewart. 2020. "Comment: Summarizing income mobility with multiple smooth quantiles instead of parameterized means." Sociological Methodology 50(1):96-111.
#' @references Fasiolo, Matteo, Simon N. Wood, Margaux Zaffran, Raphaël Nedellec, and Yannig Goude. 2020. "Fast calibrated additive quantile regression." Journal of the American Statistical Association.
#' @export

print.quantileplot <- function(x, ...) {

  # If there is a lack of convergence, warn the user.
  convergence <- sapply(x$mqgam.out$fit, function(each_fit) each_fit$converged)
  if (any(!convergence)) {
    print("Warning: Convergence not achieved for the following quantile curves.")
    print(names(convergence[!convergence]))
  }

  # Print the plot
  plot(x)
}


R/scatter.quantileplot.R
#' Create a Scatter for a Quantile Plot
#'
#' @description Converts a quantile plot into an analogous scatter plot.
#' @param object An object of class \code{quantileplot}, created by a call to \code{quantileplot()}.
#' @param fraction Numeric scalar in (0,1]. Fraction of points to be randomly sampled for inclusion in the scatter plot. Useful in very large data sets.
#' @param ... Other arguments
#'
#' @return A ggplot2 object containing a scatter plot.
#'
#' @references Lundberg, Ian, Robin C. Lee, and Brandon M. Stewart. 2021. "The quantile plot: A visualization for bivariate population relationships." Working paper.
#' @references Lundberg, Ian, and Brandon M. Stewart. 2020. "Comment: Summarizing income mobility with multiple smooth quantiles instead of parameterized means." Sociological Methodology 50(1):96-111.
#' @references Fasiolo, Matteo, Simon N. Wood, Margaux Zaffran, Raphaël Nedellec, and Yannig Goude. 2020. "Fast calibrated additive quantile regression." Journal of the American Statistical Association.
#'
#' @export
#' @import dplyr
#'
#' @examples
#' x <- rbeta(1000,1,2)
#' y <- log(1 + 9 * x) * rbeta(1000, 1, 2)
#' data <- data.frame(x = x, y = y)
#' qp <- quantileplot(y ~ s(x), data)
#' scatter.quantileplot(qp)

scatter.quantileplot <- function(object, fraction = 1, ...) {
  if (class(object) != "quantileplot") {
    stop("Argument error: The first argument object must be an object of class quantileplot")
  }
  if (fraction <= 0 | fraction > 1) {
    stop("Argument error: fraction should be a scalar greater than 0 and less than or equal to 1.")
  }
  # Initialize objects for non-standard evaluation
  x <- y <- weight <- NULL
  # Extract the plot
  for_scatter <- object$plot
  # Keep only the plot area (remove curves and densities)
  for_scatter$layers <- for_scatter$layers[(length(object$plot$layers) - 2):length(object$plot$layers)]
  # Extract the x and y variable names
  y_str <- all.vars(object$arguments$formula)[1]
  x_str <- all.vars(object$arguments$formula)[2]
  # Create a data frame for the scatter plot
  scatter_data <- data.frame(x = object$arguments$data[[x_str]],
                             y = object$arguments$data[[y_str]])
  # Append weights
  if (is.null(object$arguments$weights)) {
    scatter_data$weight <- 1
  } else {
    scatter_data$weight <- object$arguments$data[[object$arguments$weights]]
  }
  # Restrict to the *_data_range
  if (!is.null(object$arguments$x_data_range)) {
    scatter_data <- scatter_data %>%
      dplyr::filter(x > object$arguments$x_data_range[1] & x < object$arguments$x_data_range[2])
  }
  if (!is.null(object$arguments$y_data_range)) {
    scatter_data <- scatter_data %>%
      dplyr::filter(y > object$arguments$y_data_range[1] & y < object$arguments$y_data_range[2])
  }
  if (fraction < 1) {
    scatter_data <- scatter_data %>%
      dplyr::slice_sample(prop = fraction)
  }
  # Create the scatter plot
  scatter <- for_scatter +
    ggplot2::geom_point(data = scatter_data,
                        ggplot2::aes(x = x, y = y, size = weight),
                        color = "gray") +
    ggplot2::scale_size_continuous(range = c(.1,1)) +
    ggplot2::theme(legend.position = "none")
  return(scatter)
}


R/summary.quantileplot.R
#' Summary function for \code{quantileplot} objects
#' @description Summarizes the S3 class object returned by the quantileplot function
#' @param object An object of class \code{quantileplot}, which results from a call to the function \code{quantileplot}
#' @param ... Other arguments to \code{summary} commands
#' @return Prints a summary of the estimates.
#' @references Lundberg, Ian, Robin C. Lee, and Brandon M. Stewart. 2021. "The quantile plot: A visualization for bivariate population relationships." Working paper.
#' @references Lundberg, Ian, and Brandon M. Stewart. 2020. "Comment: Summarizing income mobility with multiple smooth quantiles instead of parameterized means." Sociological Methodology 50(1):96-111.
#' @references Fasiolo, Matteo, Simon N. Wood, Margaux Zaffran, Raphaël Nedellec, and Yannig Goude. 2020. "Fast calibrated additive quantile regression." Journal of the American Statistical Association.
#' @export

summary.quantileplot <- function(object, ...) {

  # Return the call to the user
  cat("Call:\n")
  print(object$call)

  # Say a bit about the plot
  cat("\nA quantileplot with the following elements:\n")
  cat("\n1. A horizontal density for the predictor\n")
  cat(paste0("\n2. Vertical densities for the conditional outcome distribution at predictor values:\n   ",
             paste0(format(object$slice_x_values, digits = 3), collapse = ", "),"\n"))
  cat(paste0("\n3. Curves for conditional quantiles of the outcome given the predictor, at quantiles:\n   ",
             paste0(names(object$mqgam.out$fit), collapse = ", "),"\n"))

  cat(paste0("\nDensities are estimated by a Gaussian product kernel with horizontal bandwidth ",
            format(object$densities$x_bw, digits = 3),
            " and vertical bandwidth ",
            format(object$densities$y_bw, digits = 3),
            ".\n"))

  # Summarize convergence of the mqgam
  converged <- sapply(object$mqgam.out$fit, function(each_fit) each_fit$converged)
  if (all(converged)) {
    cat("\nConvergence achieved for every quantile curve.\n")
  }
  if (any(!converged)) {
    cat("\nWarning: convergence not achieved for the following quantile curves.\n")
    print(names(converged[!converged]))
    cat("\nIt is possible that you have requested a very extreme quantile for which there is very little data.\n")
    cat("See the mqgam function in the qgam package to address this problem.")
  }

  # Print the plot
  if (object$arguments$uncertainty_draws) {
    print(object$with_uncertainty_draws)
  } else {
    print(object$plot)
  }
}


R/gen_curves.R
#' Generate Quantile Curves for a \code{quantileplot}
#'
#' @description Estimate smooth curves for quantiles of the outcome as a function of the predictor. This function calls \code{mqgam} from the \code{qgam} package. This function is typically called indirectly via a user call to \code{\link{quantileplot}}.
#' @param formula A bivariate model formula (e.g. \code{y ~ s(x)})
#' @param data Data frame containing the variables in \code{formula}. If \code{weights} are specified, they must be a column of \code{data}.
#' @param weights String name for sampling weights, which are a column of \code{data}. If not given, a simple random sample is assumed.
#' @param quantiles Numeric vector containing quantiles to be estimated. Values should be between 0 and 1.
#' @param show_ci Logical, defaults to \code{FALSE}. Whether to show credible intervals for the estimated smooth quantile curves.
#' @param credibility_level Numeric probability value for credible intervals; default to 0.95 to produce 95 percent credible intervals. Only relevant if \code{show_ci = TRUE}.
#' @param uncertainty_draws A whole number. If non-null, the number of simulated posterior draws to estimate for each smooth quantile curve. When used with the \code{plot} function, these appear in panels below the main plot.
#' @param inverse_transformation A function of a scalar argument. Only used in the rare use case where the outcome has an extremely skewed distribution and the user wants to estimate the quantile curves on a transformed outcome, to be brought back to the original scale for the visualization. In that case, this argument is the function to convert from the transformed outcome back to the original scale. For instance, if the outcome in the model formula is \code{log(y + 1)} then the inverse transformation should be \code{function(y) exp(y) - 1}. This is a rare use case because it is only relevant when a transformation of the outcome aids the estimation of quantile curves. If you want to visualize on a transformed scale, you should instead create a transformed variable in \code{data} rather than conducting the transformation within the model formula. For common transformations (e.g. log(y)), the \code{inverse_transformation} argument can left \code{NULL} and will be determined automatically.
#' @param second_formula Model formula to allow the learning rate to change as a function of the predictor. This is passed to \code{mqgam} as the second element in the \code{form} argument. Defaults to the same specification as \code{formula} but without the outcome variable.
#' @param argGam Additional arguments to the GAM for model fitting. Passed to mqgam.
#' @param ... Other arguments passed to \code{mqgam}.
#' @return A list of length 2. Element \code{curves} is a data frame containing the data for plotting smooth curves for quantiles of the outcome given the predictor. Element \code{mqgam.out} is the fitted object from \code{mqgam}.
#' @references Lundberg, Ian, Robin C. Lee, and Brandon M. Stewart. 2021. "The quantile plot: A visualization for bivariate population relationships." Working paper.
#' @references Lundberg, Ian, and Brandon M. Stewart. 2020. "Comment: Summarizing income mobility with multiple smooth quantiles instead of parameterized means." Sociological Methodology 50(1):96-111.
#' @references Fasiolo, Matteo, Simon N. Wood, Margaux Zaffran, Raphaël Nedellec, and Yannig Goude. 2020. "Fast calibrated additive quantile regression." Journal of the American Statistical Association.
#'
#' @export
#' @import dplyr
gen_curves <- function(formula,
                       data,
                       weights = NULL,
                       quantiles = c(.1, .25, .5, .75, .9),
                       show_ci = FALSE,
                       credibility_level = 0.95,
                       uncertainty_draws = 10,
                       inverse_transformation = NULL,
                       second_formula,
                       argGam = NULL,
                       ...) {
  # Initialize objects that will be called by non-standard evaluation.
  i <- curve <- x <- estimate <- se <- NULL

  # Extract the x variable name from the formula
  x_str <- all.vars(formula)[2]

  # See if any smooth terms used in the formula
  contains_smooth_terms <- any(grepl("s[(]",as.character(formula)))

  # Extract any transformations on the y variable
  outcome_with_transformation <- as.character(formula)[2]
  outcome_variable <- all.vars(formula)[1]
  # Determine the inverse of the transformation on the y-variable.
  # We will use this to return estimated curves to the untransformed y-scale.
  if (is.null(inverse_transformation)) {
    if (outcome_with_transformation == outcome_variable) {
      inverse_transformation <- identity
    } else if (outcome_with_transformation == paste0("log(",outcome_variable,")")) {
      inverse_transformation <- exp
    } else if (outcome_with_transformation == paste0("exp(",outcome_variable,")")) {
      inverse_transformation <- log
    } else if (outcome_with_transformation == paste0("sqrt(",outcome_variable,")")) {
      inverse_transformation <- function(x) x ^ 2
    } else {
      stop(paste("Your model formula uses a transformation",outcome_with_transformation,". This transformation will be used for fitting quantile curves, but you need to provide the inverse_transformation argument so that we can return the plot on the original scale."))
    }
  }

  # Estimate the smooth quantile curves
  if (is.null(weights) & contains_smooth_terms) {
    mqgam.out <- qgam::mqgam(list(formula, second_formula),
                             data = data, qu = quantiles,
                             argGam = argGam,
                             ...)
  } else if (!is.null(weights) & contains_smooth_terms) {
    mqgam.out <- qgam::mqgam(list(formula, second_formula),
                             data = data, qu = quantiles,
                             # When passing in the weights, normalize them to sum to the number of observations
                             argGam = c(argGam, list(weights = data[[weights]] / mean(data[[weights]]))),
                             ...)
  } else if (is.null(weights) & !contains_smooth_terms) {
    mqgam.out <- qgam::mqgam(formula,
                             data = data, qu = quantiles,
                             argGam = argGam,
                             ...)
  } else if (!is.null(weights) & !contains_smooth_terms) {
    mqgam.out <- qgam::mqgam(formula,
                             data = data, qu = quantiles,
                             argGam = c(argGam, list(weights = data[[weights]] / mean(data[[weights]]))),
                             ...)
  }

  # Split the x range into 200 points where we will make predictions
  to_predict <- data.frame(x = seq(min(data %>% group_by() %>% dplyr::select_at(x_str)),
                                   max(data %>% group_by() %>% dplyr::select_at(x_str)),
                                   length.out = 200)) %>%
    rename_with(.fn = function(x) x_str)

  # Generate point estimate and simulated curves
  quantile_curves_list <- lapply(quantiles, function(q) {
    quantile_index <- which(quantiles == q)
    # Get the model matrix on the expanded basis
    if (!contains_smooth_terms) {
      X_basis_expansion <- mgcv::predict.gam(mqgam.out$fit[[1]],
                                             type = "lpmatrix",
                                             newdata = to_predict)
    } else {
      # For smooth terms, need a different function to predict the model matrix
      # cbind adds the intercept to the smooth
      X_basis_expansion <- cbind(1,mgcv::PredictMat(mqgam.out$smooth[[1]],
                                                    data = to_predict))
    }
    # Extract the fitted coefficients beta
    beta <- mqgam.out$fit[[quantile_index]]$coefficients
    # Extract the variance-covariance matrix Sigma
    rV <- mqgam.out$fit[[quantile_index]]$rV
    sig2 <- mqgam.out$fit[[quantile_index]]$sig2
    Sigma <- rV %*% t(rV) * sig2

    # Calculate the point estimate and standard error
    yhat_point <- X_basis_expansion %*% beta
    yhat_se <- sqrt(diag(X_basis_expansion %*% Sigma %*% t(X_basis_expansion)))
    # Create a data frame of the point estimate
    point_df <- data.frame(curve = "point",
                           x = to_predict[[1]],
                           estimate = inverse_transformation(yhat_point),
                           ci.min = inverse_transformation(yhat_point - stats::qnorm((1 + credibility_level) / 2) * yhat_se),
                           ci.max = inverse_transformation(yhat_point + stats::qnorm((1 + credibility_level) / 2) * yhat_se))

    # Calculate uncertainty_draws simulated curves
    if (!is.null(uncertainty_draws)) {
      beta_star <- t(mvtnorm::rmvnorm(uncertainty_draws, mean = beta, sigma = Sigma))
      yhat_star <- inverse_transformation(X_basis_expansion %*% beta_star)
      # Create a data frame of the simulated curves
      sim_curves_df <- data.frame(yhat_star) %>%
        dplyr::rename_all(.funs = function(x) gsub("X","sim",x)) %>%
        dplyr::mutate(x = to_predict[[1]]) %>%
        reshape2::melt(id = "x", variable.name = "curve", value.name = "estimate") %>%
        dplyr::select(curve, x, estimate) %>%
        mutate(ci.min = NA, ci.max = NA)
    }
    to_return <- point_df
    if (!is.null(uncertainty_draws)) {
      to_return <- to_return  %>%
        dplyr::bind_rows(sim_curves_df)
    }
    to_return <- to_return %>%
      dplyr::mutate(Form = "Smooth",
                    Summary = "quantiles",
                    percentile = paste0(100 * q,"th"),
                    percentile_num = 100 * q)

    return(to_return)
  })
  # Convert the result to a data frame
  quantile_curves_df <- do.call(rbind, quantile_curves_list)

  return(list(curves = quantile_curves_df,
              mqgam.out = mqgam.out))
}

